lib/pbench/server/database/models/datasets.py

```diff
@@ -407,14 +406,24 @@ class Dataset(Database.Base):
 
     __tablename__ = "datasets"
 
+    # This dict defines the allowed dataset state transitions through its
+    # lifecycle. We have a set of "-ING" action states while the dataset is
+    # being mutated, and a set of "-ED" states designating the last successful
+    # mutation, from UPLOAD through INDEX and eventually DELETE.
+    #
+    # All "-ING" action states allow self-transitions, which provides for a
+    # failed action to be retried later.
+    #
+    # DELETED does not appear as a key here: we define it for completion, but
+    # (A) if a dataset entered DELETED state it could never leave and (B) on
+    # completion of the dataset delete action the dataset's SQL row has been
+    # deleted so we never actually set it to "DELETED" state.
```

Comment:
I would omit reason (A):  a user _should_ be able to upload a dataset, delete it, and then upload it again.  This would (if we retained some notion of a dataset after it was deleted) result in a transition _out of_ the `DELETED` state.  So, I think the proper answer is (B):  once a dataset enters the `DELETED` state, there is no way to tell, so there is no explicit way for it to transition to another state, so there is no need to have an entry for it in the table (although, perhaps it wouldn't hurt to have one, for completeness -- as I said, it _should_ be able to transition from `DELETED` to `UPLOADING`).

Reply:
Yeah, this is awkward territory. In a way, I shouldn't even have defined `DELETED` since we never use it. If we *did* use it, you're probably right; which I suppose argues that since I *did* define it, I should allow a transition from `DELETED` to `UPLOADING`. 🤔